src/nova_tz/keyblob.rs

```diff
@@ -0,0 +1,120 @@
+use std::array::TryFromSliceError;
+use std::convert::TryFrom;
+use std::convert::TryInto;
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Can't parse key blob: {0}")]
+    BadKeyBlob(String),
+}
+
+pub enum DigestPadAlgo {
+    RsaDigestPaddingNone = 0,
+    RsaPkcs115Sha2_256 = 1,
+    RsaPssSha2_256 = 2,
+}
+
+impl TryFrom<u32> for DigestPadAlgo {
+    type Error = Error;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(DigestPadAlgo::RsaDigestPaddingNone),
+            1 => Ok(DigestPadAlgo::RsaPkcs115Sha2_256),
+            2 => Ok(DigestPadAlgo::RsaPssSha2_256),
+            _ => Err(Error::BadKeyBlob(format!(
+                "Bad digest padding algorithm {}",
+                value
+            ))),
+        }
+    }
+}
+
+pub(crate) struct KeyBlob {
+    pub modulus: Vec<u8>,
+    pub public_exponent: Vec<u8>,
+}
+
+// Note (iegor.s): values are copied from qualcomm qseecom module
+const RSA_KEY_SIZE_MAX: usize = 512 + 16;
+const RSA_IV_LENGTH: usize = 16;
+const RSA_HMAC_LENGTH: usize = 32;
+const RSA_KEY_BLOB_LENGHT: usize = 1656;
+
+fn read_u32_le(buf: &[u8]) -> Result<(u32, &[u8]), TryFromSliceError> {
+    let (val_bytes, rest) = buf.split_at(4);
+    let parsed = u32::from_le_bytes(val_bytes.try_into()?);
+    Ok((parsed, rest))
+}
+
+fn read_u8_array_dynamic(
+    buf: &[u8],
+    maxsize: usize,
+) -> Result<(Vec<u8>, &[u8]), TryFromSliceError> {
+    let (arr_bytes, rest) = buf.split_at(maxsize);
+    let (arr_len, rest) = read_u32_le(rest)?;
+    let arr = arr_bytes[..arr_len as usize].to_vec();
+    Ok((arr, rest))
+}
+
+fn read_u8_array(buf: &[u8], size: usize) -> (Vec<u8>, &[u8]) {
+    let (arr_bytes, rest) = buf.split_at(size);
+    let arr = arr_bytes.to_vec();
+    (arr, rest)
+}
+
+impl TryFrom<&[u8]> for KeyBlob {
+    type Error = Error;
+
+    fn try_from(buf: &[u8]) -> Result<Self, Error> {
+        // Key blobs are always exactly RSA_KEY_BLOB_LENGHT long, but
+        // let's tolerate extra.
+        if buf.len() < RSA_KEY_BLOB_LENGHT {
+            return Err(Error::BadKeyBlob(format!(
```

Comment:
Please rename KeyBlob to TZKeypair and then actually create a DecodeError enum for  each of the errors so we don't have all these strings for errors

Reply:
Not sure if I understood your request about errors correctly. If it's not what you meant - please describe.